compiler/AST/symbol.cpp

```diff
@@ -568,7 +569,10 @@ bool VarSymbol::isConstant() const {
 
 
 bool VarSymbol::isConstValWillNotChange() const {
-  return hasFlag(FLAG_CONST);
+  // todo: how about QUAL_CONST ?
+  return qual == QUAL_CONST_VAL         ||
+         hasFlag(FLAG_REF_TO_IMMUTABLE) ||
+         (hasFlag(FLAG_CONST) && !hasFlag(FLAG_REF_VAR));
```

Comment:
I'm not so certain of this last clause. Wouldn't it be better to add FLAG_REG_TO_IMMUTABLE in the right places instead? In particular, I'm worried that a compiler-introduced reference temporary might end up with FLAG_CONST and be a reference. But maybe FLAG_CONST does not work that way. Can you confirm that this is OK?

Reply:
@mppf my only confirmation is that this passes testing.

I think this is a really gray area because I don't think any of us knows exactly how FLAG_CONST works.

While I am not following exactly on what you are saying, one thing we can do to strengthen this is to switch from hasFlag(FLAG_REF_VAR) to isRef(). This should reduce the change of isConstValWillNotChange() returning true incorrectly. @benharsh what do you think about it?

Reply:
I'm suspicious of this one:
https://github.com/chapel-lang/chapel/blob/master/compiler/resolution/functionResolution.cpp#L5034
(because I can't explain it)

and this one:

https://github.com/chapel-lang/chapel/blob/master/compiler/resolution/functionResolution.cpp#L5034

because it might set FLAG_CONST on a reference (a Qualifier type reference)

and this one:

https://github.com/chapel-lang/chapel/blob/master/compiler/resolution/functionResolution.cpp#L5792

because the ArgSymbol might be a reference

After that I got tired of looking... I do think that switching to isRef() there is probably a good idea.

Reply:
Switching to isRef seems worth trying.

constness representation in the compiler is still a bit of a mess, so if our existing tests (much improved over the past year) still pass, then I'm content for now.